pyccel/ast/cudaext.py

```diff
@@ -174,6 +174,62 @@ def __init__(self, dim=None):
     def dim(self):
         return self._dim
 
+
+class CudaCopy(CudaNewArray):
+    """
+    Represents a call to  cuda.copy for code generation.
+
+    Parameters
+    ----------
+    arg : list, tuple, PythonList
+
+    memory_location : str
+        'host'   the newly created array is allocated on host.
+        'device' the newly created array is allocated on device.
+    
+    is_async: bool
+        Indicates whether the copy is asynchronous or not [Default value: False]
+
+    """
+    __slots__ = ('_arg','_dtype','_precision','_shape','_rank','_order','_memory_location', '_is_async')
+
+    def __init__(self, arg, memory_location, is_async=False):
+
+        if not isinstance(arg, (PythonTuple, PythonList, Variable)):
```

Comment:
I'm not really sure if we can accept all those types, for example only `Variable` has the attribute `memory_location` in you implementation something like this should be possible:
``` python
if __name__ == '__main__':
    arr = cuda.array([1,2,3,4], memory_location='device')
    cpy = cuda.copy([1,2,3,4], 'host', is_async=True)
```

Reply:
Yes you right , it will raise an error in this case. 
It would be better to restrict the copy only from existing cuda arrays for now.
Fixed in 28f3ecf8916671459508cdbd9b75fcca28a692f4